File: src/simple-tst.cpp
#include <iostream>
#include <cstdlib>
#include <fstream>
#include <vector>
#include <sstream>
#define MAX 50

using namespace std;

struct Node
{
    char data;
    unsigned isEndOfString: 1;
    Node *left, *eq, *right;
};

Node* newNode(char data) 
{
    Node* temp = new Node();
    temp->data = data;
    temp->isEndOfString = 0;
    temp->left = temp->eq = temp->right = NULL;
    return temp;
}

void insert(Node** root, const char *word) {
    if(!(*root))
        *root = newNode(*word);
    if((*word) < (*root)->data)
        insert(&((*root)->left), word);
    else if ((*word) > (*root)->data)
        insert(&((*root)->right), word);
    else {
        if(*(word+1))
            insert(&((*root)->eq), word+1);
        else
            (*root)->isEndOfString=1;
    }
}

int search(Node *root, const char *word) {
    if(!root)
        return 0;
    if (*word < (root)->data)
        return search(root->left, word);
    else if(*word > (root)->data)
        return search(root->right, word);
    else {
        if(*(word+1) == '\0')
            return root->isEndOfString;
        return search(root->eq, word+1);
    }
}

int main() {
    Node *root = NULL;
    string s;
    ifstream infile("Word-List.txt");
    string word;
    while(getline(infile, word)) {
        istringstream iss(word);
        iss >> word;
        insert(&root, word.c_str());
    }
    cout << "Enter string to search: ";
    cin >> s;
    if(search(root, s.c_str()))
        cout << s << " found in TST\n";
    else
        cout << s << " not found in TST\n";
    return 0;
}

File: src/simple_tst.py
from random import shuffle
import cPickle as pickle
import redis

class Node:
    def __init__(self, data):
        self.data = data
        self.is_end_of_str = False
        self.left = None
        self.eq = None
        self.right = None


def insert(root, word, i=0):
    if root is None:
        root =  Node(word[i])
    if word[i] < root.data:
        root.left = insert(root.left, word, i)
    elif word[i] > root.data:
        root.right = insert(root.right, word, i)
    else:
        if i+1 < len(word):
            root.eq = insert(root.eq, word, i+1)
        else:
            root.is_end_of_str = True
    return root


def search(root, word, i=0):
    if root is None:
        return False
    if word[i] < root.data:
        return search(root.left, word, i)
    elif word[i] > root.data:
        return search(root.right, word, i)
    else:
        if i+1 == len(word):
            return root.is_end_of_str
        return search(root.eq, word, i+1)

def get_root():
    root = None
    with open("Word-List.txt") as f:
        words  = f.readlines()
        shuffle(words)
        for word in words:
            root = insert(root, word.strip())
    return root

# r = redis.StrictRedis(host='localhost', port=6379, db=0)
# pickled_obj = pickle.dumps(root)
# r.set('tst', pickled_obj)
# root = pickle.loads(r.get('tst'))


